### ISSUE TITLE
How does border_mode work?

### ISSUE THREAD

**Issue by jasper95** [fchollet/keras#1984]
I'm confused what to use, valid or same. I also don't know what does it do. I can't find it in the Docs

**Comment by FighterLYL** [comment]
The `valid` means there is no padding around input or feature map, while `same` means there are some padding around input or feature map, making the output feature map's size same as the input's

**Comment by jasper95** [comment]
@FighterLYL thanks.

**Comment by jasper95** [comment]
@FighterLYL  i saw an [example](https://github.com/fchollet/keras/blob/master/examples/cifar10_cnn.py#L43-L65) that doesn't have model.add(ZeroPadding) but doing border_mode='same' automatically adds zero padding?

**Comment by FighterLYL** [comment]
Yes. But what i said above is on the condition that `stride=1`, for other cases, refer to the code:
```
def conv_output_length(input_length, filter_size, border_mode, stride):
    if input_length is None:
        return None
    assert border_mode in {'same', 'valid'}
    if border_mode == 'same':
        output_length = input_length
    elif border_mode == 'valid':
        output_length = input_length - filter_size + 1
    return (output_length + stride - 1) // stride
```

**Comment by ashrafatef** [comment]
what is value of zero padding in border_mode="valid" ?

**Comment by alyato** [comment]
@FighterLYL  I don't understand  when the stride isn't 1.
I think that `the input == the output ` whatever the stride is set   when `the valid_mode = same`

**Comment by abdulqayyum** [comment]
this is good discussion. I want to know that if i add zeroPadding2D >1 then should I use border mode same or valid? because i have already added padding. Or it is indication that zero Padding layer is added so border_mode should be valid?

**Comment by SarthakYadav** [comment]
@abdulqayyum  Adding zeropadding is not necessarily an indicator that your input has been aptly padded so as to get a same sized input. 

Zero padding with `(F-1)/2` zeros, where F is the size of your (square) filter at the Convolution layer will result in getting a same sized output. 
Therefore, if your filters are of dimension 3x3, then choosing a padding of 1 will result in dimension preservation of your input data.

You can either apply ZeroPadding as per your requirements, followed by the Convolutional Layer, or let the Convolutional layer do the job by simply specifying border_mode= 'same'  if you intend to preserve the dimensions of your input.


@fchollet  I think this issue can be closed now.

**Comment by MichaelMa2014** [comment]
May I ask what the default value for `border_mode` is?

**Comment by Danielhiversen** [comment]
@MichaelMa2014 Default is valid:
https://github.com/fchollet/keras/blob/master/keras/layers/convolutional.py#L425

**Comment by MichaelMa2014** [comment]
Thank you @Danielhiversen ! 

And for anyone interested, `border_mode` is now called `padding`.

**Comment by apacha** [comment]
@FighterLYL, I think your explanation should go into the [official documentation](https://keras.io/layers/convolutional/) because it is currently lacking any information apart from the two possible options, or did I miss something?

**Comment by yagyapandeya** [comment]
full means apply filter wherever it partly overlaps with the input, i.e. padding of dimension (filter rows - 1, filter columns - 1)
half means that there is a padding of (filter rows // 2 rows, filter columns // 2 columns)
same means there are some padding around input or feature map, making the output feature map's size same as the input's.
valid means there is no padding around input or feature map, i.e. padding of dimension (0,0).
